platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java

```diff
@@ -13,204 +13,259 @@
-    if (felixFileOriginal != null) {
+  @Override
+  public void handleDelete(String pid) throws IOException {
+    CachedConfigData cachedConfigData = pidDataMap.remove(pid);
+    File fileFromCache = (cachedConfigData != null) ? cachedConfigData.getFelixFile() : null;
+    if (fileFromCache != null && fileFromCache.exists()) {
       LOGGER.debug("Deleting file because config was deleted for pid {}", pid);
       try {
-        performOperation(configActions.delete(felixFileOriginal.toPath()));
-      } catch (IllegalArgumentException | ConfiguratorException e) {
+        Files.delete(fileFromCache.toPath());
+        SECURITY_LOGGER.info("Removing a deleted config: {}", fileFromCache.toPath());
+      } catch (IOException e) {
         LOGGER.debug("Problem deleting config file: ", e);
+        // Synchronous with config admin, so we can report the failure to the UI this way
+        throw e;
       }
     }
   }
 
-  private void performOperation(Operation operation) {
-    Configurator configurator = configuratorFactory.getConfigurator();
-    configurator.add(operation);
-    OperationReport report = configurator.commit("Synchronizing the etc directory");
-    if (report.containsFailedResults()) {
-      LOGGER.error("One or more file operations failed, see debug logs for more info");
-      if (LOGGER.isDebugEnabled()) {
-        report
-            .getFailedResults()
-            .stream()
-            .map(Result::getError)
-            .filter(Optional::isPresent)
-            .map(Optional::get)
-            .map(Exception.class::cast)
-            .forEach(e -> LOGGER.debug("Failed file operation: ", e));
+  /**
+   * In this block, based upon the circumstances of where it gets called, there's a guarantee that
+   * {@link ConfigurationContext#getConfigFile()} will not return null.
+   */
+  private void writeIfNecessary(ConfigurationContext context, CachedConfigData data)
+      throws IOException {
+    File dest = context.getConfigFile();
+    if (dest.exists()) {
+      // Improvement: These would become checksums
+      Dictionary<String, Object> fileState = data.getProps();
+      Dictionary<String, Object> configState = context.getSanitizedProperties();
+      processPasswords(context.getServicePid(), configState, encryptionService::encryptValue);
+      if (!equalDictionaries(configState, fileState)) {
+        try (OutputStream outputStream = new FileOutputStream(dest)) {
+          getAppropriateStrategy(dest.toPath()).write(outputStream, configState);
+        }
+        SECURITY_LOGGER.info("Updating config file: {}", dest.toPath());
+        data.setProps(configState);
       }
     }
   }
 
-  private Configuration getConfiguration(ConfigurationEvent configurationEvent) {
-    try {
-      return configurationAdmin.getConfiguration(configurationEvent.getPid(), null);
-    } catch (IOException e) {
-      LOGGER.debug("Problem reading from config admin: ", e);
-      return null;
+  private void processPasswords(
```

Comment:
✏️ An alternate (but ultimately not much better) streaming implementation:
```java
    ObjectClassDefinition objectClassDefinition = ddfConfigAdmin.getObjectClassDefinition(pid);
    if (objectClassDefinition == null) {
      return;
    }
    
    Stream.of(objectClassDefinition)
        .map(ocd -> ocd.getAttributeDefinitions(ObjectClassDefinition.ALL))
        .flatMap(Arrays::stream)
        .filter(ad -> ad.getType() == AttributeDefinition.PASSWORD)
        .map(AttributeDefinition::getID)
        .filter(id -> dictionary.get(id) instanceof String)
        .forEach(id -> dictionary.put(id, passwordTransform.apply((String) dictionary.get(id))));
```

Reply:
Ah good. I'll explore this. ❓ Could I replace the `flatMap(Arrays::stream)` call with a regular `map` call? 

I came up with something similar but my approach won't work because I can't get the key back for overwriting the dictionary entry:
```
private void processPasswordsStreaming(String pid,
          Dictionary<String, Object> dictionary,
          Function<String, String> passwordTransform) {
    ObjectClassDefinition objectClassDefinition = ddfConfigAdmin.getObjectClassDefinition(pid);
    if (objectClassDefinition != null) {
      Arrays.stream(objectClassDefinition.getAttributeDefinitions(ObjectClassDefinition.ALL))
              .filter(attribute -> attribute.getType() == AttributeDefinition.PASSWORD)
              .map(AttributeDefinition::getID)
              .map(dictionary::get)
              .filter(String.class::isInstance)
              .map(String.class::cast)
              .map(passwordTransform)
              .forEach(dictionary::put); // Compile error - don't have key
    }
  }
```

Reply:
I think the answer to my own question is that converting the singleton stream of `objectClassDefinition` to `AttributeDefinition[]` does not explode the array's internal elements to the stream itself.

Reply:
@coyotesqrl The one thing I don't like about your stream proposal is that it ends up hashing the dictionary twice. This is one of the many reasons why I hate the old interfaces like Dictionary. The new Map would allow to do a compute() which would hash once the dictionary in search of the entry and allow you to change it if yo u want.

@Lambeaux It is preferable to use Stream.of() instead of Arrays.stream()

Reply:
I think an iterative solution is probably for the best here, to be honest. The Dictionary and the fact that they're using Arrays for return/input values everywhere make Streaming solutions less efficient.